File: leetcodecpp/leetcode/OneZeroOne.cpp
//
// Created by John on 2020/2/22.
//

#include "OneZeroOne.h"
#include "NineSevenNine.h"
#include <queue>

using namespace std;

class Solution {
public:
//    bool isSymmetric(TreeNode * root) {
//        if (root == NULL) {
//            return true;
//        }
//        return traverse(root, root);
//    }
    bool isSymmetric(TreeNode * root) {
        queue<TreeNode *> q;
        q.push(root);
        q.push(root);
        while (!q.empty()) {
            TreeNode * t1 = q.front();
            TreeNode * t2 = q.front();
            if (t1 == NULL && t2 == NULL) {
                continue;
            }
            if (t1 == NULL || t2 == NULL) {
                return false;
            }
            if (t1->val != t2->val) {
                return false;
            }
            q.push(t1->left);
            q.push(t2->right);
            q.push(t1->right);
            q.push(t2->left);
        }
        return true;
    }
private:
    bool traverse(TreeNode * left, TreeNode * right) {
        if (left == NULL && right == NULL) {
            return true;
        }
        if (left == NULL || right == NULL) {
            return false;
        }
        return left->val == right->val && traverse(left->left, right->right) && traverse(left->right, right->left);
    }
};
//int main() {
//    TreeNode * lefttwo = new TreeNode(2);
//    TreeNode * leftthree = new TreeNode(3);
//    TreeNode * righttwo = new TreeNode(2);
//    TreeNode * rightthree = new TreeNode(3);
//    righttwo->right = rightthree;
//    lefttwo->right = leftthree;
//    TreeNode * one = new TreeNode(1);
//    one->left = lefttwo;
//    one->right = righttwo;
//
//    Solution * solution = new Solution();
//    solution->isSymmetric(one);
//
//}

File: leetcodejava/src/main/java/com/leetcode/OneZeroOne.java
package com.leetcode;

import com.entity.TreeNode;

import java.util.LinkedList;
import java.util.Queue;

/**
 * @author: John
 * @date: 2020-05-19 21:03
 * @description: 101
 **/
public class OneZeroOne {

    public boolean isSymmetric(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        q.add(root);
        while (!q.isEmpty()) {
            TreeNode t1 = q.poll();
            TreeNode t2 = q.poll();
            if (t1 == null && t2 == null) {
                continue;
            }
            if (t1 == null || t2 == null) {
                return false;
            }
            if (!t1.val.equals(t2.val)) {
                return false;
            }
            q.add(t1.left);
            q.add(t2.right);
            q.add(t1.right);
            q.add(t2.left);
        }
        return true;
    }


    /**
     * recursive
     */
//    public boolean isSymmetric(TreeNode root) {
//        return isMirror(root, root);
//    }
//
//    private boolean isMirror(TreeNode t1, TreeNode t2) {
//        if (t1 == null && t2 == null) {
//            return true;
//        }
//        if (t1 == null || t2 == null) {
//            return false;
//        }
//        return (t1.val.equals(t2.val)) && isMirror(t1.right, t2.left)
//        && isMirror(t1.left, t2.right);
//    }
}


